history-network.md

```diff
@@ -0,0 +1,300 @@
+
+The `node-id` is a 32-byte identifier. We define the `distance` function that maps a pair of `node-id` values to a 256-bit unsigned integer identically to the Discovery v5 network.
+
+```
+distance(n1, n2) = n1 XOR n2
+```
+
+Similarly, we define a `logdistance` function identically to the Discovery v5 network.
+
+```
+logdistance(n1, n2) = log2(distance(n1, n2))
+```
+
+### Content: Keys and Values
+
+The chain history DHT stores the following data items:
+
+* Block headers
+* Block bodies
+* Receipts
+
+Each of these data items are represented as a key-value pair. Denote the key for a data item by `content-key`. Denote the value for an item as `content`.
+
+We derive a `content-id` from the `content-key` as `H(content-key)` where `H` denotes the [placeholder] hash function, which outputs 32-byte values. The `content-id` represents the key in the DHT that we use for `distance` calculations.
+
+All `content-key` values are encoded and decoded according to SSZ sedes.
+
+#### Block Header
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
+content-type = 0x01
+```
+
+#### Block Body
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
+content-type = 0x02
+```
+
+#### Receipts
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
+content-type = 0x03
+```
+
+### Radius
+
+We define a `distance` function that maps a `node-id` and `content-id` pair to a 256-bit unsigned integer identically to the `distance` function for pairs of `node-id` values.
+
+Each node specifies a `radius` value, a 256-bit unsigned integer that represents the data that a node is "interested" in.
+
+```
+interested(node, content) = distance(node.id, content.id) <= node.radius
+```
+
+A node is expected to maintain `radius` information for each node in its local node table. A node's `radius` value may fluctuate as the contents of its local key-value store change.
+
+### Protocol Messages
+
+All protocol messages are transmitted via the `TALKREQ` and `TALKRESP` messages defined by the Discovery v5 protocol.
+
+Below is the general message encoding for all `TALKREQ` messages.
+
+```
+protocol              = portal-history
+request               = protocol-message-type || protocol-message
+protocol-message-type = uint8
+protocol-message      = byte array
+```
+
+Below is the general message encoding for all `TALKRESP` messages.
+
+```
+response              = protocol-message-type || protocol-message
+protocol-message-type = uint8
+protocol-message      = byte array
+```
+
+All `protocol-message` values are encoded and decoded according to SSZ sedes.
+
+If the size of a message would exceed the max packet size of 1280 bytes, then the transmission of the message contents would occur over [uTP](https://github.com/ethereum/stateless-ethereum-specs/blob/master/discv5-utp.md).
```

Comment:
I think the wording here may lead to some confusion.  One way of reading this would be:

> I want to send a FOUNDNODES message.  The total size of the ENR records exceeds 1280 bytes.  So the FOUNDNODES request would be sent over uTP.

I think you can move this down to the STORE message discription since I think that is the only place where uTP will be used.

Reply:
Looking at this again, the uTP would also be used in a `FoundContent` message if the data payload is sufficiently large, so I think we can leave it here and specify that it only applies to the transmission of content.

Reply:
Ah yes, I forgot about FOUNDCONTENT.